### ISSUE TITLE
Documentation and Examples for BertPretrainer(V2)

### ISSUE THREAD

**Issue by areiner-novetta** [tensorflow/models#9699]
# Prerequisites

Please answer the following question for yourself before submitting an issue.

- [+] I checked to make sure that this issue has not been filed already.

## 1. The entire URL of the documentation with the issue

https://github.com/tensorflow/models/blob/master/official/nlp/modeling/models/bert_pretrainer.py

## 2. Describe the issue

I'm attempting to use BertPretrainer(V2) and the MaskedLM layer it produces, but there are no examples for it and no docs about it.  The bert_pretrainer_test.py shows how to plug an example through the model, but doesn't have any of the critical information about batching, training loops, etc.

My primary problem:  I figured out that the MaskedLM layer expects a *list of positions*, but that list itself will be ragged, and I don't see a good way to use it with batching, besides trying to group together inputs of similar size with the same number of mask tokens.  Additionally, I then have to extract the same positions from my target labels to apply the loss function -- not hard, but tedious and probably unnecessary given a good example somewhere.  

I'm tempted to create my own custom version that gets all predictions but simply zeros out losses in non-masked slots just before averaging them.  But that is tedious too and I don't want to go down that route unless I have to.  Am I missing something?  Are there any public projects using this that I could look at?

**Comment by saberkun** [comment]
BertPretrainerV2 is used in https://github.com/tensorflow/models/blob/master/official/nlp/tasks/masked_lm.py and we have corresponding unit tests for it.

For your problems, (1) sorry, I don't think the implementation support ragged inputs of positions. In fact, we only tested with fixed shape tensors; (2) In the masked_lm.py, we are doing masking and only use the loss of tokens that have weights >0. It probably can help you.

**Comment by areiner-novetta** [comment]
Thanks for the speedy response!  My impression from reading through this is that it's relatively new and some features may not be full fleshed out.  In particular, all examples appear to use batch-size of 1 and num_token_predictions=20.  In my particular application, both these numbers will be variable and implementation complexity will ramp up quickly.

I feel like the major problem I have is the MLM mask position vectors when using batches that have different numbers of masked tokens.

- My tokenized input word IDs start in a ragged tensor, which I can rectangularize through `input_ragged.to_tensor(default_value=tokenizer.encode('[PAD]')`.  No problem here.
- However, I cannot do the same with the ragged MLM position vectors, as there is no good value for padding.  If I pick an index out of range, it causes errors.  If I pick an index in range, it will amplify the losses for that token dramatically.  

The only two workarounds I can think of are:
- Use batch size = 1 (seems unreasonable)
- Group inputs by size to batch up only those with the same number of masked tokens (more reasonable).

Unless I missing some underlying motivation/justification, I would propose MLM positions would instead be another mask.  The mask would be the same size as inputs, and only have 1s for the 15% of tokens that are to be predicted.  It could then be directly broadcast over the model outputs to zero the non-masked values in the loss computation.  This is potentially slightly slower than the way it is implemented now, but probably negligible compared to the rest of the computations during training.  It might be simpler in implementation than all the gather/gather_indexes calls.

**Comment by saberkun** [comment]
Yes, you have good thoughts. First of all, the code works on TPU usually needs to be relatively static shape. The dynamic operations are not well compiled/optimized.
Second, I do see your approach with ragged inputs in other gpu libraries.
Lastly, the input processing script we have now generated fixed number of max masked tokens with padding.
Right now, we are gathering a fixed number of positions and mask the pad tokens in the loss.

**Comment by areiner-novetta** [comment]
Thanks for the feedback.  I will close this issue for now but will be keeping an eye on this code for further developments.  At least you linked me to that concrete example from the `tasks` directory.  Thanks.
